src/ZippedArrays.jl

```diff
@@ -96,26 +101,44 @@ Use the syntax `Z.args` to retrieve the arrays `A`, `B`, `C`, etc.
 #     L = number of zipped arrays
 #     I = indexing style (true for linear, false otherwise)
 #     S = type of L-tuple of zipped arrays
-struct ZippedArray{T,N,L,I,S<:ArrayTuple{L,N}} <: AbstractArray{T,N}
+struct ZippedArray{T,N,L,I,S<:Union{ArrayTuple{L,N},ArrayNamedTuple{L,N}}} <: AbstractArray{T,N}
     args::S
 end
 
 const at_least_one_array_to_zip = ArgumentError("there must be at least one array to zip")
 const not_same_ndims = DimensionMismatch("arrays to zip must have the same number of dimensions")
 
-ZippedArray() = throw(at_least_one_array_to_zip)
 ZippedArray(args::AbstractArray...) = ZippedArray(args)
 ZippedArray(args::Tuple{Vararg{AbstractArray}}) = throw(not_same_ndims)
 function ZippedArray(args::S) where {L,N,S<:ArrayTuple{L,N}}
+    L==0 && throw(at_least_one_array_to_zip)
+    T = Tuple{map(eltype, args)...}
+    I = get_index_style(args...) === IndexLinear()
+    return ZippedArray{T,N,L,I,S}(args)
+end
+
+ZippedArray(;args...) = ZippedArray(values(args))
+function ZippedArray(args::S) where {L,N,S<:ArrayNamedTuple{L,N}}
+    L==0 && throw(at_least_one_array_to_zip)
     T = Tuple{map(eltype, args)...}
     I = get_index_style(args...) === IndexLinear()
     return ZippedArray{T,N,L,I,S}(args)
 end
 
-ZippedArray{T}() where {T} = throw(at_least_one_array_to_zip)
 ZippedArray{T}(args::AbstractArray...) where {T} = ZippedArray{T}(args)
 ZippedArray{T}(args::Tuple{Vararg{AbstractArray}}) where {T} = throw(not_same_ndims)
 function ZippedArray{T}(args::S) where {T,L,N,S<:ArrayTuple{L,N}}
+    L==0 && throw(at_least_one_array_to_zip)
+    # We do not enforce that the number of arrays matches the number of fields
+    # is `T` is a structure type to let the caller provides its own builder.
+    !(T <: Tuple) || destruct_count(T) == L || throw(ArgumentError(
+        "number of entries in element type `$T` must match number of arguments"))
+    I = get_index_style(args...) === IndexLinear()
+    return ZippedArray{T,N,L,I,S}(args)
+end
+ZippedArray{T}(;args...) where {T} = ZippedArray{T}(values(args))
+function ZippedArray{T}(args::S) where {T,L,N,S<:ArrayNamedTuple{L,N}}
```

Comment:
I had to duplicate this code because Julia was complaining about type ambiguities if I define `ZippedArray{T}(args::S) where {T,L,N,S<:Union{ArrayTuple{L,N},ArrayNamedTuple{L,N}}}`

Reply:
I like the idea very much. Thank you Ferréol!